app: Urbackup

FeatherKing opened the pull request:
in its current state it passes my various test cases but i still cant connect to the UI. This app is a bit particular about UDP broadcasts so I am going to test on some other machines before i request review.

FeatherKing commented:
tests are passing now on my linux system. Ran into some snags testing on apple silicon.

The only logo image i can find for Urbackup is this one https://github.com/uroni/urbackup_backend/blob/dev/urbackupserver/www/images/urbackup.png?raw=true

looking at the contributing guide, i will need to have someone put that on the truenas CDN and give me another link.

This is my first contribution, please let me know what else is required.

FeatherKing commented:
@stavros-k is there a way to generate a host port? or a list of host ports used? i see many in the 30xxx range but im not seeing how they are generated. For example in Urbackup i just used the default 55414 which is pretty high, but i wasnt clear if there is a list somewhere where i should reserve a port.

stavros-k commented on ix-dev/community/urbackup/templates/docker-compose.yaml:
```diff
@@ -0,0 +1,33 @@
+{% do c1.healthcheck.set_test("http", {"port": values.network.web_port}) %}
+
+
+{% if values.urbackup.zfs_enabled %}
+  {% do c1.devices.add_device("/dev/zfs", "/dev/zfs") %}
+{% endif %}
```
Hello,
With this here, this PR is not going to be merged.
if you think it is still of value to add it without this. Let me know.

FeatherKing replied:
ill remove it. im currently running this on a non-truenas system with a zfs pool and this option disabled. I was just adapting what the author had in [his repo here](https://github.com/uroni/urbackup-server-docker)

FeatherKing replied:
fixed

stavros-k commented on ix-dev/community/urbackup/templates/docker-compose.yaml:
```diff
@@ -0,0 +1,33 @@
+  {% do c1.devices.add_device("/dev/zfs", "/dev/zfs") %}
+{% endif %}
+
+{% if values.urbackup.btrfs_enabled %}
+  {% do c1.add_caps(["SYS_ADMIN"]) %}
+{% endif %}
```
Truenas system does not even have btrfs. So not sure about this either.

FeatherKing replied:
ill remove

FeatherKing replied:
fixed

FeatherKing commented:
ive addressed the feedback and made the default to **not** be host networking. There are some discovery conveniences that you lose, but it works fine without host networking (this is how i run it at home). Host networking can be enabled though and both of those tests are passing.

stavros-k commented on ix-dev/community/urbackup/templates/library/base_v2_1_9/__init__.py:
Please delete the older library directory.

FeatherKing replied:
removed

stavros-k commented on ix-dev/community/urbackup/migrations/migrate_from_kubernetes:
delete migrations please.

FeatherKing replied:
removed

stavros-k commented on ix-dev/community/urbackup/templates/test_values/basic-values.yaml:
```diff
@@ -0,0 +1,53 @@
+    bind_mode: published
+    port_number: 35623
+
+run_as:
+  user: 0
+  group: 0
```
Why testing as root?

FeatherKing replied:
i cant seem to make it work without root. I tested as userid 568 and added all docker capabilities but it fails trying to lock /etc/passwd

> 2025-01-24 10:04:15 urbackup-server-1  | usermod: Permission denied.
2025-01-24 10:04:15 urbackup-server-1  | usermod: cannot lock /etc/passwd; try again later.

on my non truenas server i run this application and inside the container it is setup as root

> [media@nas ~]$ docker exec -it urbackup-server-1 /bin/bash
root@cafeebfd16f5:/# id
uid=0(root) gid=0(root) groups=0(root)
root@cafeebfd16f5:/#

maybe i can mount /etc/passwd as read only. ill try

FeatherKing replied:
mount /etc/passwd read only got me further

> 2025-01-24 10:16:29 urbackup-server-1  | usermod: user 'urbackup' does not exist

not sure about this one without modifying the dockerfile upstream

FeatherKing replied:
https://github.com/uroni/urbackup-server-docker/blob/latest/entrypoint.sh

> Giving the user and group "urbackup" the provided UID/GID
if [[ $PUID != "" ]]
then
	usermod -u $PUID -o urbackup
else
	usermod -u 101 -o urbackup
fi
if [[ $PGID != "" ]]
then
	groupmod -g $PGID -o urbackup
else
	groupmod -g 101 -o urbackup
fi
chown urbackup:urbackup /backups
chown urbackup:urbackup /var/urbackup
exec urbackupsrv "$@"

it wants to run usermod but the user doesnt exist.. hm.

stavros-k replied:
Okay so, it runs as root but starts subprocesses as non-root.

So,
c1.set_user(0, 0)
c1.add_caps(["CHOWN", "FOWNER", "DAC_OVERRIDE", "SETUID", "SETGID"])

You still leave the run_as section as is, and it will automagically set the PUID etc.

FeatherKing replied:
great. updated to include this

FeatherKing commented:
Addressed the latest feedback. I also tried to incorporate some of the changes you made to my ESPHome PR. Mainly, adding in the permissions container logic.

stavros-k commented on ix-dev/community/urbackup/ix_values.yaml:
```diff
@@ -0,0 +1,14 @@
+  internal_backup_port_3: 35623
+  backup_path: /backups
+  database_path: /var/urbackup
+  www_path: /usr/share/urbackup
```
what does the wwwfolder contain?

FeatherKing replied:
<img width="532" alt="image" src="https://github.com/user-attachments/assets/85ae014e-d2e0-478b-a6e7-d8e38fc61094" />

Im not sure why someone would want to bind mount it but i added that option from [here](https://github.com/uroni/urbackup-server-docker/blob/52023db293ddc9a55c00e17b06b49a76e48d11af/README.md?plain=1#L50)

i run this at home and have never mounted that dir. i can remove and maybe people can optionally mount it

stavros-k replied:
Yes let's remove it. It's probably for modifying the webui.

centauri commented:
I don't know if I'm supposed to write here, but urbackup in it's current config on scale does not work. In "host" it does.
It just hangs at "Downloading dataplan database..." and no client will add.

And shouldn't ZFS be enabled?

```
2025-01-30 20:54:24.817525+00:00usermod: no changes
2025-01-30 20:54:24.855070+00:00Raising nice-ceiling to 35 failed. (errno=1)
2025-01-30 20:54:24.856110+00:002025-01-30 21:54:24: Starting HTTP-Server on port 55414
2025-01-30 20:54:24.856202+00:002025-01-30 21:54:24: HTTP: Server started up successfully!
2025-01-30 20:54:24.856667+00:002025-01-30 21:54:24: Generating Server identity...
2025-01-30 20:54:24.856924+00:002025-01-30 21:54:24: Generating Server private/public ECDSA key...
2025-01-30 20:54:24.857788+00:002025-01-30 21:54:24: Calculating public key...
2025-01-30 20:54:25.070118+00:002025-01-30 21:54:25: Generating Server private/public DSA key...
2025-01-30 20:54:25.119299+00:002025-01-30 21:54:25: Calculating public key...
2025-01-30 20:54:25.138297+00:002025-01-30 21:54:25: Generating Server token...
2025-01-30 20:54:25.141679+00:002025-01-30 21:54:25: WARNING: Upgrading...
2025-01-30 20:54:25.142028+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 8
2025-01-30 20:54:25.168858+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 9
2025-01-30 20:54:25.186036+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 10
2025-01-30 20:54:25.191160+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 11
2025-01-30 20:54:25.195885+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 12
2025-01-30 20:54:25.200241+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 13
2025-01-30 20:54:25.204938+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 14
2025-01-30 20:54:25.209421+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 15
2025-01-30 20:54:25.216497+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 16
2025-01-30 20:54:25.223419+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 17
2025-01-30 20:54:25.229887+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 18
2025-01-30 20:54:25.238618+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 19
2025-01-30 20:54:25.246777+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 20
2025-01-30 20:54:25.258651+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 21
2025-01-30 20:54:25.269995+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 22
2025-01-30 20:54:25.278028+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 23
2025-01-30 20:54:25.285855+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 24
2025-01-30 20:54:25.293798+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 25
2025-01-30 20:54:25.302028+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 26
2025-01-30 20:54:25.309399+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 27
2025-01-30 20:54:25.316812+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 28
2025-01-30 20:54:25.328275+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 29
2025-01-30 20:54:25.337009+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 30
2025-01-30 20:54:25.348353+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 31
2025-01-30 20:54:25.359886+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 32
2025-01-30 20:54:25.369288+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 33
2025-01-30 20:54:25.377334+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 34
2025-01-30 20:54:25.385752+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 35
2025-01-30 20:54:25.393731+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 36
2025-01-30 20:54:25.429663+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 37
2025-01-30 20:54:25.438789+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 38
2025-01-30 20:54:25.447425+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 39
2025-01-30 20:54:25.456420+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 40
2025-01-30 20:54:25.465592+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 41
2025-01-30 20:54:25.477098+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 42
2025-01-30 20:54:25.485740+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 43
2025-01-30 20:54:25.497424+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 44
2025-01-30 20:54:25.507200+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 45
2025-01-30 20:54:25.533956+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 46
2025-01-30 20:54:25.547643+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 48
2025-01-30 20:54:25.560310+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 49
2025-01-30 20:54:25.570084+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 50
2025-01-30 20:54:25.577896+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 51
2025-01-30 20:54:25.588594+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 52
2025-01-30 20:54:25.597077+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 53
2025-01-30 20:54:25.607887+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 54
2025-01-30 20:54:25.617448+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 55
2025-01-30 20:54:25.626236+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 56
2025-01-30 20:54:25.631362+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 57
2025-01-30 20:54:25.637197+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 58
2025-01-30 20:54:25.642716+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 59
2025-01-30 20:54:25.647763+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 60
2025-01-30 20:54:25.655036+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 61
2025-01-30 20:54:25.662569+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 62
2025-01-30 20:54:25.668759+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 63
2025-01-30 20:54:25.674679+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 64
2025-01-30 20:54:25.683322+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 65
2025-01-30 20:54:25.688964+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 66
2025-01-30 20:54:25.694555+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 67
2025-01-30 20:54:25.700167+00:002025-01-30 21:54:25: WARNING: Upgrading database to version 68
2025-01-30 20:54:25.708632+00:002025-01-30 21:54:25: WARNING: Done.
2025-01-30 20:54:25.732870+00:002025-01-30 21:54:25: Deleting database journal...
2025-01-30 20:54:25.736614+00:002025-01-30 21:54:25: Copying/reflinking database...
2025-01-30 20:54:25.736853+00:002025-01-30 21:54:25: Reflink ioctl failed. errno=11
2025-01-30 20:54:25.738271+00:002025-01-30 21:54:25: WARNING: Creating file entry index. This might take a while...
2025-01-30 20:54:25.738327+00:002025-01-30 21:54:25: Getting number of files...
2025-01-30 20:54:25.738452+00:002025-01-30 21:54:25: Dropping index...
2025-01-30 20:54:25.738776+00:002025-01-30 21:54:25: Starting creating files index...
2025-01-30 20:54:25.739246+00:002025-01-30 21:54:25: Creating backupid index...
2025-01-30 20:54:25.739608+00:002025-01-30 21:54:25: Syncing...
2025-01-30 20:54:25.741075+00:002025-01-30 21:54:25: Renaming back result...
2025-01-30 20:54:25.746038+00:002025-01-30 21:54:25: SQLite: recovered 1 frames from WAL file /var/urbackup/backup_server.db-wal code: 283
2025-01-30 20:54:25.747479+00:002025-01-30 21:54:25: SQLite: recovered 2 frames from WAL file /var/urbackup/backup_server_settings.db-wal code: 283
2025-01-30 20:54:25.751652+00:002025-01-30 21:54:25: Started UrBackup...
2025-01-30 20:54:25.754879+00:002025-01-30 21:54:25: Removing temporary files...
2025-01-30 20:54:25.755406+00:002025-01-30 21:54:25: Recreating temporary folder...
2025-01-30 20:54:25.761922+00:00Cannot become root user
2025-01-30 20:54:25.767351+00:002025-01-30 21:54:25: Image mounting disabled: Cannot become root user
2025-01-30 20:54:25.772907+00:00Cannot become root user
2025-01-30 20:54:25.773563+00:002025-01-30 21:54:25: Backup destination cannot handle subvolumes and snapshots. Snapshots disabled.
2025-01-30 20:54:25.777569+00:002025-01-30 21:54:25: Reflink ioctl failed. errno=11
2025-01-30 20:54:25.807738+00:002025-01-30 21:54:25: Broadcasting on ipv4 interface eth0 addr 172.16.3.2
2025-01-30 20:54:26.253073+00:002025-01-30 21:54:26: InternetService: Server started up successfully!
2025-01-30 20:54:26.258561+00:002025-01-30 21:54:26: UrBackup Server start up complete.
2025-01-30 20:54:26.258836+00:002025-01-30 21:54:26: Looking for old Sessions... 0 sessions
2025-01-30 20:54:26.258925+00:002025-01-30 21:54:26: Server started up successfully!
2025-01-30 20:54:27.260763+00:002025-01-30 21:54:27: Downloading version file...
2025-01-30 20:54:27.308027+00:002025-01-30 21:54:27: Downloading signature...
2025-01-30 20:54:27.355051+00:002025-01-30 21:54:27: Downloading old signature...
2025-01-30 20:54:27.405950+00:002025-01-30 21:54:27: Getting update file URL...
2025-01-30 20:54:27.451308+00:002025-01-30 21:54:27: Downloading update file...
2025-01-30 20:56:30.050330+00:002025-01-30 21:56:30: Successfully downloaded update file.
2025-01-30 20:56:30.050610+00:002025-01-30 21:56:30: Downloading version file...
2025-01-30 20:56:30.096966+00:002025-01-30 21:56:30: Downloading signature...
2025-01-30 20:56:30.143123+00:002025-01-30 21:56:30: Getting update file URL...
2025-01-30 20:56:30.189589+00:002025-01-30 21:56:30: Downloading update file...
2025-01-30 20:56:35.195702+00:002025-01-30 21:56:35: Successfully downloaded update file.
2025-01-30 20:56:35.195936+00:002025-01-30 21:56:35: Downloading server version info...
2025-01-30 20:56:35.241568+00:002025-01-30 21:56:35: Downloading dataplan database...
```

FeatherKing commented:
@centauri can you open an issue? but i think this is working as designed. the container without host networking is limited to the docker network so it cant discover clients. you would need to add hints in the urbackup UI. host networking worked when i tested in my local setup.

also the logs do typically just end there at "downloading dataplan database"
